[SPARK-6287][MESOS] Add dynamic allocation to the coarse-grained Mesos scheduler

dragos opened the pull request:
This is largely based on extracting the dynamic allocation parts from @tnachen's #3861.

dragos committed:
Add dynamic allocation to the Spark coarse-grained scheduler.

AmplabJenkins commented:
Can one of the admins verify this patch?

dragos commented:
See #4990 for the second part of the original PR (external shuffle service)

tnachen commented:
@pwendell @andrewor14

sryza commented:
One thing to watch out for here is that, `ExecutorAllocationManager` calculates the number of cores per executor based on the YARN default.  If this is different than the Mesos default, then something probably needs to be changed there.

andrewor14 commented:
@dragos is this ready for review? Can you rebase to master first? I'm assuming this depends logically on your other patch #4990.

AmplabJenkins commented:
Can one of the admins verify this patch?

andrewor14 commented:
@dragos is this still targeted for 1.4?

dragos commented:
Yes, I'd love to get this in in 1.4. I'll push a rebased version.

andrewor14 commented:
Thanks, but it still has conflicts

dragos commented:
Sorry, I didn't get to it.

dragos commented:
Unfortunately conflicts were really non-trivial and it took me quite a bit of time to bring this up to date. I guess it's too late for 1.4, but I'd like to push this forward as soon as possible. I don't want to go through another round fixing conflicts. :) /cc @tnachen

tnachen commented:
@dragos is this rebased now? seems like 1.4 branch is cut at this point, let me ask @andrewor14

dragos commented:
It's rebased. It can (still) be merged without conflicts :)

andrewor14 commented:
ok to test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32052 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32052/consoleFull) for   PR 4984 at commit [`a68c3bf`](https://github.com/apache/spark/commit/a68c3bf7fb9621bb9792e605aaf795b7867cbbba).

SparkQA commented:
[Test build #32052 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32052/consoleFull) for   PR 4984 at commit [`a68c3bf`](https://github.com/apache/spark/commit/a68c3bf7fb9621bb9792e605aaf795b7867cbbba).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32052/
Test FAILed.

tnachen commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -19,6 +19,9 @@ package org.apache.spark.scheduler.cluster.mesos
 import java.util.{Collections, List => JList}
+import java.util.concurrent.locks.ReentrantLock
+
+import com.google.common.collect.HashBiMap
```
This should be after scala imports

tnachen commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -130,7 +142,7 @@ private[spark] class CoarseMesosSchedulerBackend(
           .format(prefixEnv, runScript) +
-        s" --driver-url $driverUrl" +
+        s" --driver-url $driverURL" +
         s" --executor-id ${offer.getSlaveId.getValue}" +
```
After testing this patch with a real Mesos cluster I got a bunch of errors when I kill an executor, start it again and try to do recompute a shuffle job such as:

15/05/10 18:15:55 WARN TaskSetManager: Lost task 11.0 in stage 3.0 (TID 53, 10.70.15.58): FetchFailed(null, shuffleId=0, mapId=-1, reduceId=11, message=
org.apache.spark.shuffle.MetadataFetchFailedException: Missing an output location for shuffle 0
	at org.apache.spark.MapOutputTracker$$anonfun$org$apache$spark$MapOutputTracker$$convertMapStatuses$1.apply(MapOutputTracker.scala:389)
	at org.apache.spark.MapOutputTracker$$anonfun$org$apache$spark$MapOutputTracker$$convertMapStatuses$1.apply(MapOutputTracker.scala:386)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:108)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:244)
	at scala.collection.mutable.ArrayOps$ofRef.map(ArrayOps.scala:108)
	at org.apache.spark.MapOutputTracker$.org$apache$spark$MapOutputTracker$$convertMapStatuses(MapOutputTracker.scala:385)
	at org.apache.spark.MapOutputTracker.getServerStatuses(MapOutputTracker.scala:172)
	at org.apache.spark.shuffle.hash.BlockStoreShuffleFetcher$.fetch(BlockStoreShuffleFetcher.scala:42)
	at org.apache.spark.shuffle.hash.HashShuffleReader.read(HashShuffleReader.scala:40)
	at org.apache.spark.rdd.ShuffledRDD.compute(ShuffledRDD.scala:90)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:277)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:244)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:63)
	at org.apache.spark.scheduler.Task.run(Task.scala:70)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:213)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:722)
)

I hit another similiar error in the past when I worked on this feature, and Aaron Davidson told me that for external shuffle service to work every new executor that is launched must have a new executor ID and shouldn't use the same one as before, therefore in my closed PR (3681) I introduced a new sparkExecutorId method that concats slave id and task id to form a new executor id.
I think you need to also make every new executor id that is launched unique so the external shuffle client/service interaction works, and not just simply use slave id.

dragos replied:
I couldn't reproduce this one yet, but the assertion failure happens often indeed.

tnachen replied:
How is your Mesos cluster setup look like?

dragos replied:
3 Mesos slaves in the local network, driver running on my machine. I modified the code to use the task ID, as in your original PR.

I'm running spark-shell with something like this:

```
val rdd = sc.parallelize(1 to 10000000, 500)

val rdd1 = rdd.zipWithIndex.groupBy(_._1 / 100)

rdd1.cache()

rdd1.collect()
```

spark-defaults.conf:

```
spark.mesos.coarse				true
spark.shuffle.service.enabled	true

spark.dynamicAllocation.enabled				true
spark.dynamicAllocation.minExecutors 			1
spark.dynamicAllocation.maxExecutors 			3
spark.dynamicAllocation.executorIdleTimeout		15
```

I can see the executors being shut down and coming back to life...

tnachen replied:
Ah ok, I wasn't sure if you tested it before you changed it or not. With new executor ID it should be right.

dragos replied:
Yes, but the problem's that I couldn't reproduce it *before* the change.. :(

tnachen commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -281,4 +319,40 @@ private[spark] class CoarseMesosSchedulerBackend(
+      }
+    }
+
+    assert(pendingRemovedSlaveIds.size <= taskIdToSlaveId.size)
```
This assert is failing for me when I let the ExecutorAllocationManager start killing idling executors.
Adding some print lines I realize that the assumption that these two lists are exclusive is wrong:

15/05/10 18:36:11 WARN CoarseMesosSchedulerBackend: pending: Set(20150509-065852-3381634314-5050-1960-S2, 20150509-065852-3381634314-5050-1960-S0, 20150509-065852-3381634314-5050-1960-S1)
15/05/10 18:36:11 WARN CoarseMesosSchedulerBackend: current: {2=20150509-065852-3381634314-5050-1960-S2}

When we kill the Mesos task we add to the pending list, but it's still in the current list until the task status is returned, so it can easily get into the situation where pending > current. We should remove this assert, and also make the executorLimitOption to be:
Option(Math.max(0, taskIdToSlaveId.size - pendingRemovedSlaveIds.size))

dragos commented:
Thanks @tnachen for your comments. I'll look into them ASAP.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32397 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32397/consoleFull) for   PR 4984 at commit [`4053462`](https://github.com/apache/spark/commit/4053462bd49ffb240cae5191f1c226b6265d7ce9).

tnachen commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -281,4 +319,38 @@ private[spark] class CoarseMesosSchedulerBackend(
+    // executors, that means the scheduler wants to set the limit that is less than
+    // the amount of the executors that has been launched. Therefore, we take the existing
+    // amount of executors launched and deduct the executors killed as the new limit.
+    executorLimitOption = Option(taskIdToSlaveId.size - pendingRemovedSlaveIds.size)
```
Would this make the limit negative now?

dragos replied:
Oh, I forgot to push my latest changes. Yes, it's `max` now, as you suggested

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32400 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32400/consoleFull) for   PR 4984 at commit [`4e831e2`](https://github.com/apache/spark/commit/4e831e27f48ce0f8cc36002af556330518fcc249).

SparkQA commented:
[Test build #32397 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32397/consoleFull) for   PR 4984 at commit [`4053462`](https://github.com/apache/spark/commit/4053462bd49ffb240cae5191f1c226b6265d7ce9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32397/
Test FAILed.

dragos commented:
Hm, the test fails because it can't find the mesos driver in the classpath. Note that the call is triggered when trying to mock `MesosSchedulerDriver`, which loads the native library in its [static initializer](https://github.com/apache/mesos/blame/master/src/java/src/org/apache/mesos/MesosSchedulerDriver.java#L61). Not sure if this changed recently, but if so.. I'm not sure how to fix it.

```
sbt.ForkMain$ForkError: no mesos in java.library.path
	at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1886)
	at java.lang.Runtime.loadLibrary0(Runtime.java:849)
	at java.lang.System.loadLibrary(System.java:1088)
	at org.apache.mesos.MesosNativeLibrary.load(MesosNativeLibrary.java:54)
	at org.apache.mesos.MesosNativeLibrary.load(MesosNativeLibrary.java:79)
	at org.apache.mesos.MesosSchedulerDriver.<clinit>(MesosSchedulerDriver.java:61)
	at sun.reflect.GeneratedSerializationConstructorAccessor1261.newInstance(Unknown Source)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
	at org.objenesis.instantiator.sun.SunReflectionFactoryInstantiator.newInstance(SunReflectionFactoryInstantiator.java:56)
	at org.objenesis.ObjenesisBase.newInstance(ObjenesisBase.java:73)
	at org.mockito.internal.creation.jmock.ClassImposterizer.createProxy(ClassImposterizer.java:111)
	at org.mockito.internal.creation.jmock.ClassImposterizer.imposterise(ClassImposterizer.java:51)
	at org.mockito.internal.util.MockUtil.createMock(MockUtil.java:52)
	at org.mockito.internal.MockitoCore.mock(MockitoCore.java:41)
	at org.mockito.Mockito.mock(Mockito.java:1014)
	at org.mockito.Mockito.mock(Mockito.java:909)
	at org.scalatest.mock.MockitoSugar$class.mock(MockitoSugar.scala:74)
	at org.apache.spark.scheduler.mesos.CoarseMesosSchedulerBackendSuite.mock(CoarseMesosSchedulerBackendSuite.scala:45)
	at org.apache.spark.scheduler.mesos.CoarseMesosSchedulerBackendSuite$$anonfun$2.apply$mcV$sp(CoarseMesosSchedulerBackendSuite.scala:92)
	at org.apache.spark.scheduler.mesos.CoarseMesosSchedulerBackendSuite$$anonfun$2.apply(CoarseMesosSchedulerBackendSuite.scala:91)
	at org.apache.spark.scheduler.mesos.CoarseMesosSchedulerBackendSuite$$anonfun$2.apply(CoarseMesosSchedulerBackendSuite.scala:91)
```

tnachen commented:
Interesting, we shouldn't actually try to load the library since it might not be present in their machine for everyone. I wonder how come no one reported this before though, I doubt everyone installed Mesos library already. Let me take a deeper look to see if we can do something about it

tnachen commented:
Oh nothing should have changed too.

SparkQA commented:
[Test build #32400 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32400/consoleFull) for   PR 4984 at commit [`4e831e2`](https://github.com/apache/spark/commit/4e831e27f48ce0f8cc36002af556330518fcc249).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32400/
Test FAILed.

dragos commented:
This might be a side-effect of moving to mockito from EasyMock, per [this comment](https://github.com/apache/spark/pull/3861#discussion_r24644511).

dragos commented:
Scratch that. @tnachen can you have a look again? I switched to using the `SchedulerDriver` interface (I had to modify the field type in `MesosSchedulerUtils`, but it seems cleaner this way. Now we don't use the concrete `MesosSchedulerDriver` in mocked tests so we are fine.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32504 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32504/consoleFull) for   PR 4984 at commit [`064abf4`](https://github.com/apache/spark/commit/064abf46b4544d9b9912430a3a24e588ef6652c8).

SparkQA commented:
[Test build #32504 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32504/consoleFull) for   PR 4984 at commit [`064abf4`](https://github.com/apache/spark/commit/064abf46b4544d9b9912430a3a24e588ef6652c8).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32504/
Test PASSed.

dragos commented:
I tested this on a 3-node Mesos cluster (as before) and everything looked fine:

spark-defaults.conf:

```
spark.mesos.coarse				true
spark.shuffle.service.enabled	true

spark.dynamicAllocation.enabled				true
spark.dynamicAllocation.minExecutors 			1
spark.dynamicAllocation.maxExecutors 			3
spark.dynamicAllocation.executorIdleTimeout		15
```


in spark-shell:
```
val rdd = sc.parallelize(1 to 10000000, 500)

val rdd1 = rdd.zipWithIndex.groupBy(_._1 / 100)

rdd1.cache()

rdd1.collect()
```

andrewor14 commented:
Hi all, I had a discussion with @pwendell offline about this patch. At this point it would seem too late to put it into the 1.4 release since it is a new feature that hasn't gone through a ton of review from the community or committers. I think it would be a great addition to 1.5 after spending some more time testing it.

dragos commented:
@andrewor14 thanks for the heads up. I understand it might be late for 1.4. Is there anything we can do to help? Maybe @tnachen can report on his experience with the latest fixes.

I totally understand that 1.4 is the highest priority, but it'd be great to get some feedback so the PR doesn't bit rot again. Again, let me know if there's anything I can do to speed this up.

andrewor14 commented:
Yeah, in the mean time I will continue my reviews on this. I also want to have this feature in soon, but we should make sure that it's at least well-tested before we merge it.

tnachen commented:
I verified that the assertion is gone but I'm still getting the old error that I saw before:

Caused by: java.lang.RuntimeException: java.lang.RuntimeException: Failed to open file: /tmp/spark-5e4149e8-6d2a-4863-9d62-8e9c0a8f37fd/blockmgr-13dcb8c8-4abc-4d96-9554-af367ee9f05d/35/shuffle_0_5_0.index
	at org.apache.spark.network.shuffle.ExternalShuffleBlockResolver.getSortBasedShuffleBlockData(ExternalShuffleBlockResolver.java:202)
	at org.apache.spark.network.shuffle.ExternalShuffleBlockResolver.getBlockData(ExternalShuffleBlockResolver.java:112)
	at org.apache.spark.network.shuffle.ExternalShuffleBlockHandler.receive(ExternalShuffleBlockHandler.java:74)
	at org.apache.spark.network.server.TransportRequestHandler.processRpcRequest(TransportRequestHandler.java:114)
	at org.apache.spark.network.server.TransportRequestHandler.handle(TransportRequestHandler.java:87)
	at org.apache.spark.network.server.TransportChannelHandler.channelRead0(TransportChannelHandler.java:101)
	at org.apache.spark.network.server.TransportChannelHandler.channelRead0(TransportChannelHandler.java:51)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:333)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:319)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:254)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:333)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:319)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:333)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:319)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:163)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:333)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:319)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:787)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:130)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:511)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354)
	at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:116)
	at java.lang.Thread.run(Thread.java:722)
Caused by: java.io.FileNotFoundException: /tmp/spark-5e4149e8-6d2a-4863-9d62-8e9c0a8f37fd/blockmgr-13dcb8c8-4abc-4d96-9554-af367ee9f05d/35/shuffle_0_5_0.index (No such file or directory)
	at java.io.FileInputStream.open(Native Method)
	at java.io.FileInputStream.<init>(FileInputStream.java:138)
	at org.apache.spark.network.shuffle.ExternalShuffleBlockResolver.getSortBasedShuffleBlockData(ExternalShuffleBlockResolver.java:191)
	... 26 more

I think we need to figure out more what's going on, and I was wondering why @dragos never see the same issues I had.

andrewor14 commented:
@tnachen do you have a small reproduction of what you ran to trigger the exception? What configs did you set?

tnachen commented:
Here is the spark settings:
spark.executor.uri=http://10.165.143.201:8000/spark-1.4.0-SNAPSHOT-bin-2.2.0.tgz
spark.mesos.coarse=true
spark.dynamicAllocation.enabled=true
spark.shuffle.service.enabled=true

Simply shuffle job:
val rdd = sc.parallelize(0 until 10000).map((_,1)).reduceByKey(_ + _)
rdd.count()

I'm running the shuffle service daemon on each slave using Marathon.

dragos commented:
Probably not what's happening here, but I had issues when I ran the shuffle service under a different user than Mesos, that didn't have permissions to read shuffle files written by executors. However, here it's a `FileNotFoundException`....

dragos commented:
Ok, I could reproduce the failure. Indeed slaves miss the `/tmp/spark-*` directories, investigating who's deleting them. Thanks for catching this!

dragos commented:
It's only a hunch so far, but I think that default containerization in Mesos is deleting all files written by Spark executors (under `/tmp`) as soon as the executor is killed. At least, the directory disappears the moment the Mesos task is killed.

I'll see if I can use the [shared filesystem isolator][cont] to map `/tmp` between the host and Spark executors.

[cont]: http://mesos.apache.org/documentation/latest/mesos-containerizer/

dragos commented:
The shuffle files are placed in a temporary directory that is deleted on exit by a shutdown hook in the `Utils` class. It would work as long as the JVM that spawned the executor does not exit, but that's not the case on Mesos.

dragos committed:
Add persistent shuffle directory

Call “applicationRemoved” when the application ends and delete external shuffle files.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33468 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33468/consoleFull) for   PR 4984 at commit [`19920d3`](https://github.com/apache/spark/commit/19920d33cec48e329e7b4498dea0498060c253cd).

SparkQA commented:
[Test build #33468 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33468/consoleFull) for   PR 4984 at commit [`19920d3`](https://github.com/apache/spark/commit/19920d33cec48e329e7b4498dea0498060c253cd).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33468/
Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33469 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33469/consoleFull) for   PR 4984 at commit [`fdbbc32`](https://github.com/apache/spark/commit/fdbbc323b78c9fc6e39285640f04b863a4298091).

SparkQA commented:
[Test build #33469 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33469/consoleFull) for   PR 4984 at commit [`fdbbc32`](https://github.com/apache/spark/commit/fdbbc323b78c9fc6e39285640f04b863a4298091).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33469/
Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33470 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33470/consoleFull) for   PR 4984 at commit [`d50249a`](https://github.com/apache/spark/commit/d50249a3d592f8f9c72ffac6f0a1570b2409c641).

dragos commented:
For easier testing, I uploaded a [pre-built binary here](http://downloads.typesafe.com/tmp/spark-prs/spark-1.4.0-SNAPSHOT-bin-dynalloc013-fdbbc32.tgz)

The gist is the following:

- shuffle files are stored in the temporary directory (`/tmp/blockmgr/...`), but not deleted on exit
- shuffle files are now deleted by the external shuffle service when the application exits (this functionality was already in the external shuffle service, but never called)
- the BlockManagerMaster keeps an ever-growing list of BlockManagerIds, essentially remembering all executors that ever lived
- hooked the `applicationRemoved` call on BlockManager.stop (only the driver will call this method). Now it goes through all known executors and calls the external shuffle service telling it to remove the application and delete shuffle files

I verified that this works in the previous scenario and that files are correctly deleted when the application exits. I'd really appreciate some feedback on this approach/tests.

A couple of things I would still look into
- [ ] make `applicationRemoved` async so all messages are sent in parallel when stopping the block manager
- [ ] check that there are no duplicate messages (a certain worker may wake up again to host a new executor, but there's only one shuffle service on that host)

SparkQA commented:
[Test build #33470 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33470/consoleFull) for   PR 4984 at commit [`d50249a`](https://github.com/apache/spark/commit/d50249a3d592f8f9c72ffac6f0a1570b2409c641).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class RemoveApplication extends BlockTransferMessage `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33470/
Test PASSed.

dragos commented:
@andrewor14, I had a great chat with @tnachen yesterday and went through the code of this PR. There are some subtle interactions. I'd be very happy to jump on a hangout with you to answer any questions you may have. I'd really like to push this in time for 1.5.

I tried to organize the code a bit, so the "applicationRemoved" feature is self-contained and introduced in the last commit. It should be easy to review separately.

andrewor14 commented:
Hi @dragos I will take another pass over this within the next day or two. Sorry for slipping.

dragos commented:
No worries. When I chatted to @tnachen there were a few misunderstandings w.r.t to who deletes the shuffle files. Here's the story in a nutshell:

There are three levels on the file system:
  - configured local dirs (this is what the user can specify in config)
    - spark *local root dir*. This is created by Spark to hold all temp files, and is deleted by a shutdown hook on exit). This is level is **missing** in YARN and Standalone
      - shuffle dir. This is **not** deleted by Spark on exit when the external shuffle service is on.

What happens on Mesos is that the shuffle dir is indeed not deleted on exit, but it's **parent** directory is!  You won't notice that in Yarn, though, since the intermediate temporary directory is not part of the picture.

The relevant code is in `Utils.getOrCreateLocalRootDirsImpl`.

andrewor14 commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -268,8 +268,10 @@ private[spark] class ExecutorAllocationManager(
       numExecutorsToAdd = 1
-      logInfo(s"Lowering target number of executors to $numExecutorsTarget because " +
-        s"not all requests are actually needed (previously $oldNumExecutorsTarget)")
+      if (numExecutorsTarget != oldNumExecutorsTarget) {
```
this was already fixed in master. You can just rebase

andrewor14 commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -1355,10 +1354,14 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
    */
-  private[spark] def supportDynamicAllocation =
-    master.contains("yarn") || _conf.getBoolean("spark.dynamicAllocation.testing", false)
+  private[spark] def supportDynamicAllocation = {
```
not your code, but can you add return type `Boolean` here

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -59,9 +62,24 @@ private[spark] class CoarseMesosSchedulerBackend(
-  val failuresBySlaveId = new HashMap[String, Int] // How many times tasks on each slave failed
+  val taskIdToSlaveId: HashBiMap[Int, String] = HashBiMap.create[Int, String]
+  // How many times tasks on each slave failed
+  val failuresBySlaveId: HashMap[String, Int] = new HashMap[String, Int]
```
can you rename this to `slaveIdToNumFailures` to be more consistent with the other maps

andrewor14 replied:
never mind, just realized this is not your code. Feel free to leave this one as is.

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -59,9 +62,24 @@ private[spark] class CoarseMesosSchedulerBackend(
+  val failuresBySlaveId: HashMap[String, Int] = new HashMap[String, Int]
+
+  // the total number of executors we aim to have
+  private[mesos] var executorLimitOption: Option[Int] = None
```
please keep this private. I don't believe it's used anywhere outside of this class.

andrewor14 replied:
also, can you add a comment on when this is expected to be defined vs empty? If I don't use dynamic allocation it looks like this will always be `None`, is that right?

dragos replied:
It's used in tests.

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -286,4 +324,40 @@ private[spark] class CoarseMesosSchedulerBackend(
+    if (executorLimitOption.map(_ != requestedTotal).getOrElse(true)) {
+      logInfo("Capping the total amount of executors to " + requestedTotal)
+    }
+    executorLimitOption = Option(requestedTotal)
```
this will always return `Some(int)` I think. You can just use `Some` here so it's clearer at a glance. Same in L360

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -59,9 +62,24 @@ private[spark] class CoarseMesosSchedulerBackend(
+   *  Return the current executor limit, which may be [[Int.MaxValue]]
+   *  before properly initialized.
+   */
+  def executorLimit: Int = executorLimitOption.getOrElse(Int.MaxValue)
```
I think this one can be `private[mesos]`

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -59,9 +62,24 @@ private[spark] class CoarseMesosSchedulerBackend(
+   */
+  def executorLimit: Int = executorLimitOption.getOrElse(Int.MaxValue)
+
+  private[mesos] val pendingRemovedSlaveIds = new HashSet[String]
```
private

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -152,6 +164,15 @@ private[spark] class CoarseMesosSchedulerBackend(
     command.build()
   }
 
+  protected def driverURL: String = {
```
a little strange that this is protected. Can you just make it private and introduce an if case for the tests:
```
val testing = conf.get("spark.testing", "false").toBoolean
if (testing) {
  "driverURL"
} else {
  AkkaUtils.address(...)
}
```

dragos replied:
I find it a bit strange that you prefer having testing code in the normal code path, but I can do that :)

andrewor14 replied:
I just found it strange for this class to have protected methods, because it implies we're expecting it to be subclassed when it's not.

By the way can you make this private now that this doesn't get overridden anywhere?

dragos replied:
Sure thing!

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -169,14 +190,15 @@ private[spark] class CoarseMesosSchedulerBackend(
 
       for (offer <- offers) {
-        val slaveId = offer.getSlaveId.toString
+        val slaveId = offer.getSlaveId.getValue
```
what's the difference?

dragos replied:
If it even worked before, it did so by accident. You need the value of SlaveID, not some debug string that happens to be returned by `toString`

andrewor14 replied:
yeah, this looks more corect

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -169,14 +190,15 @@ private[spark] class CoarseMesosSchedulerBackend(
         val mem = getResource(offer.getResourcesList, "mem")
         val cpus = getResource(offer.getResourcesList, "cpus").toInt
-        if (totalCoresAcquired < maxCores &&
+        if (taskIdToSlaveId.size < executorLimitOption.getOrElse(Int.MaxValue) &&
```
`taskIdToSlaveId.size < executorLimit`. You already defined this up there.

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -206,8 +228,7 @@ private[spark] class CoarseMesosSchedulerBackend(
           // Filter it out
-          d.launchTasks(
-            Collections.singleton(offer.getId), Collections.emptyList[MesosTaskInfo](), filters)
+          d.declineOffer(offer.getId)
```
what's the difference? (sorry I don't know mesos that well)

dragos replied:
It's the same thing, just clearer in its intentions. According to Javadocs on [SchedulerDriver](http://mesos.apache.org/api/latest/java/org/apache/mesos/SchedulerDriver.html#launchTasks(java.util.Collection,%20java.util.Collection,%20Filters)):

> .. Invoking this function with an empty collection of tasks declines offers in their entirety (see declineOffer(OfferID, Filters)).

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -264,17 +286,33 @@ private[spark] class CoarseMesosSchedulerBackend(
     }
   }
 
+  private def sparkExecutorId(slaveId: String, taskId: String) = "%s/%s".format(slaveId, taskId)
```
Is the motivation here that each slave can now launch many executors, and we can't just use the same slave ID as the executor ID? If so we should document this clearly in a java doc.

andrewor14 replied:
also can you add return type here and everywhere else? There are many places that are missing return types.

dragos replied:
Sorry about that. The rest of the Scala world allows inferred types on private methods. What's the policy in Spark? I see that at least some `vals` use inferred types.

dragos replied:
Regarding the motivation, this part I just took from @tnachen's original PR. I think he mentioned something about executor IDs needing to be different because of the external shuffle service. @tnachen can you comment on this?

andrewor14 replied:
We have inferred types for vals everywhere, but I think we try to enforce it for even small methods.

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -286,4 +324,40 @@ private[spark] class CoarseMesosSchedulerBackend(
+
+  override def doKillExecutors(executorIds: Seq[String]): Boolean = {
+    if (mesosDriver == null) {
+      logWarning("Asked to kill executors before the executor was started.")
```
This reads funny. Did you mean "before the *mesos driver* was started?"

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -142,8 +154,8 @@ private[spark] class CoarseMesosSchedulerBackend(
-        s" --driver-url $driverUrl" +
-        s" --executor-id ${offer.getSlaveId.getValue}" +
+        s" --driver-url $driverURL" +
+        s" --executor-id ${sparkExecutorId(offer.getSlaveId.getValue, taskId.toString)}" +
```
minor: can you put this in a variable `val executorId = sparkExecutorId(...)` and use it here? For readability

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -264,17 +286,33 @@ private[spark] class CoarseMesosSchedulerBackend(
-        // Note that the slave ID corresponds to the executor ID on that slave
-        slaveIdsWithExecutors -= slaveId.getValue
-        removeExecutor(slaveId.getValue, "Mesos slave lost")
+  /** Called when a slave is lost or a Mesos task finished. Update local view on
```
minor: in Spark we use java doc style comments:
```
/**
 * Called when a slave...
 */
```

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -264,17 +286,33 @@ private[spark] class CoarseMesosSchedulerBackend(
+          removeExecutor(sparkExecutorId(slaveId, taskId.toString), reason)
+        }
+        pendingRemovedSlaveIds -= slaveId
+        slaveIdsWithExecutors -= slaveId
```
@tnachen just a heads up, after #4027 we can have multiple executors on the same slave, at which point this assumption is no longer correct.

@dragos could you add a comment here that says:
```
// TODO: This assumes one Spark executor per Mesos slave,
// which may no longer be true after SPARK-5095
```

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -264,17 +286,33 @@ private[spark] class CoarseMesosSchedulerBackend(
+   *  slave IDs that we might have asked to be killed. It also notifies the driver
+   *  that an executor was removed.
+   */
+  private def executorTerminated(d: SchedulerDriver, slaveId: String, reason: String) {
```
In Spark we require a `Unit` return type for these methods. Not all of the existing codebase already does this but we're starting to enforce it. At the very least new code should have it. (Here and many other places in this patch)

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -286,4 +324,40 @@ private[spark] class CoarseMesosSchedulerBackend(
+      val slaveId = executorId.split("/")(0)
+      if (slaveIdToTaskId.contains(slaveId)) {
+        mesosDriver.killTask(
+          TaskID.newBuilder().setValue(slaveIdToTaskId.get(slaveId).toString).build)
```
need parentheses `build()`. It's not a getter.

andrewor14 commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -286,4 +324,40 @@ private[spark] class CoarseMesosSchedulerBackend(
+    // executors, that means the scheduler wants to set the limit that is less than
+    // the amount of the executors that has been launched. Therefore, we take the existing
+    // amount of executors launched and deduct the executors killed as the new limit.
+    executorLimitOption = Option(Math.max(0, taskIdToSlaveId.size - pendingRemovedSlaveIds.size))
```
Actually, here we also assume that each slave has exactly one executor, which is problematic. Really we should be tracking the executors that are pending to be removed, not the slaves.

Note that the parent class already keeps track of this. There's already this thing called `executorsPendingToRemove` that we can just make protected. Also, this should probably take into account the executors that are pending to be added as well, right? i.e.

```
executorLimitOption = Some(Math.max(
  0, numExistingExecutors + numPendingExecutors - executorsPendingToRemove.size))
  
```

dragos replied:
The current assumption is that we never launch more than one executor per slave id. That's enforced in `resourceOffer`, last condition is `!slaveIdsWithExecutors.contains(slaveId)`. I agree it's worthwhile to relax that limitation, but let's leave that for another PR and review this one within the bounds of the current Spark state. I'm the first to acknowledge the need for a serious refactoring of this code, and I'm willing to do it. But let's make some baby steps first and get this in. :)

I believe the current logic (again, coming from @tnachen) is making too many assumptions. In fact, I would prefer to **not** modify `executorLimit` at all at this point. That's entirely the responsibility of the parent class, that implements the logic for the whole feature (and works the same in Yarn, standalone and soon, Mesos).

As a side note, my take is that this implementation should definitely not get more knowledge about the superclass. It should work entirely within the defined interface, which is `doRequestTotalExecutors` and `doKillExecutors`. They are the only (properly documented) hooks for implementors.

andrewor14 replied:
ok, fair enough. We'll let @tnachen fix this in his patch. :)

andrewor14 replied:
Ok, fair enough, we'll let @tnachen deal with this in his patch. :)

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -111,12 +113,19 @@ private[spark] class BlockManager(
-    blockTransferService
+  private[spark] val shuffleClient = mockShuffleClient.getOrElse(createShuffleClient)
+
+  private def createShuffleClient: ShuffleClient = {
```
This doesn't need to be a method. It's probably fine to just do
```
private[spark] val shuffleClient = mockShuffleClient.getOrElse {
  if (externalShuffleServiceEnabled) ...
}
```

dragos replied:
I think it's more readable this way, so unless you feel strongly about it I'd leave it as such.

andrewor14 replied:
that's fine, we can keep it...

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1223,6 +1232,9 @@ private[spark] class BlockManager(
   }
 
   def stop(): Unit = {
+    if ((blockManagerId ne null) && blockManagerId.isDriver) {
```
elsewhere in Spark we just do `if (blockManagerId != null && blockManagerId.isDriver)`

andrewor14 commented on network/shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleClient.java:
```diff
@@ -138,6 +139,26 @@ public void registerWithShuffleServer(
+   * @throws IOException
+   */
+  public void applicationRemoved(String host,
+      int port,
```
This is really strange. When I have a callback called `applicationRemoved` I would expect to see an appID, not a host and a port. All of that should be transparent to the caller and handled internally right? The other thing is that there's already similar logic in `ExternalShuffleBlockResolver#applicationRemoved`.

dragos replied:
Note that other (all?) methods in this class take a host and a port. This is not a callback. This is a client wrapper for connecting to the ExternalShuffleClient. They can be on any executor. How would you tell the client where to connect to?

dragos replied:
The `ExternalShuffleBlockResolver` is the server part. The client part that you see here is actually calling that method, through the RPC layer. This class (the client) is used by executors to connect to other executors for fetching blocks, and offers a synchronous API on top of the network. In **this** case, this client is also used by the driver to communicate to *all* executors that may have shuffle files that need to be deleted.

andrewor14 commented on core/src/test/scala/org/apache/spark/scheduler/mesos/CoarseMesosSchedulerBackendSuite.scala:
```diff
@@ -0,0 +1,187 @@
+  }
+
+  private def createSchedulerBackend(taskScheduler: TaskSchedulerImpl,
+    driver: SchedulerDriver): CoarseMesosSchedulerBackend = {
```
style:
```
private def createSchedulerBackend(
    taskScheduler: TaskSchedulerImpl,
    driver: SchedulerDriver): CoarseMesosSchedulerBackend = {
  ...
}
```

andrewor14 commented on core/src/test/scala/org/apache/spark/scheduler/mesos/CoarseMesosSchedulerBackendSuite.scala:
```diff
@@ -0,0 +1,187 @@
+
+import org.mockito.Matchers._
+import org.mockito.Mockito._
+import org.mockito.{ ArgumentCaptor, Matchers }
```
style: no space before after `{}`. Also all of these third party apps should just be grouped together. See other files.

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala:
```diff
@@ -124,10 +124,16 @@ private[spark] class DiskBlockManager(blockManager: BlockManager, conf: SparkCon
+  /**
+   * Create local directories for storing block data. These directories are
+   * located inside configured local directories and won't
+   * be deleted on JVM exit when using the external shuffle service.
```
I'm pretty confused. When external shuffle service is enabled, the executors already do not clean up the shuffle files when the JVM exits.

dragos replied:
Can you please read [my comment](https://github.com/apache/spark/pull/4984#issuecomment-117351436) again? I tried to explain to the best of my ability. If that's not clear, I'm happy to try again, but please ask more specific questions.

andrewor14 replied:
I just read it again. You mentioned that "on Mesos the shuffle dir is not deleted, but its parent directory is". I'm confused about two things: First, if the parent directory is deleted, wouldn't everything it automatically be deleted as well?

Second, I actually don't see how the parent directory (the middle layer in your comment) is deleted. Are you referring to `DiskBlockManager#doStop()`? AFAIK that only deletes the temporary directory created inside of the parent directory (i.e. the shuffle dir), and we don't do this if external shuffle service. I can't find the code that deletes the middle layer itself.

andrewor14 replied:
I just read your comment again. I still don't see how the directory layout is related to cleaning up shuffle files. The reason why we don't clean up shuffle files in Mesos (and standalone mode) is simply because the shuffle service doesn't know when an application exits. When shuffle service is enabled, [executors no longer clean up the shuffle files on exit](https://github.com/apache/spark/blob/1ce6428907b4ddcf52dbf0c86196d82ab7392442/core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala#L162), so no one cleans these files up anymore. All we need to do then is to add this missing code path.

Since the external shuffle service already [knows](https://github.com/apache/spark/blob/1ce6428907b4ddcf52dbf0c86196d82ab7392442/network/shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java#L147) about the `localDirs` on each executor, it can just go ahead and delete these directories (which contain the shuffle files written). Could you explain why the directory structure needs to change? Why is it not sufficient to just remove the shuffle directories?

dragos replied:
You are conflating two different issues.

1. Of course shuffle files were deleted! The were deleted as soon as an executor got killed. That's the reason why @tnachen reported those `FileNotFound` failures! Their parent directory is deleted on a shutdown hook (installed by `createTempDir`. That includes all subdirectories, regardless of the test you point to. Yes, it won't delete them on `doStop` (I was initially fooled by that code too), but the *parent dir* (and obviously everything underneath) *is* deleted on exit, through the shutdown hook. I verified this using log statements, so I'm pretty sure that's the case. Could we have a Hangout session to go through this? I have the feeling we're talking past each other.

So, once I fixed that, I had a new problem: these files need to be eventually cleaned up, or they would pile up indefinitely.

2. As you rightfully pointed out, the external shuffle service needs to delete those files. In my implementation I delete them when the driver stops. Therefore I send `applicationRemoved`, from the driver, to each Mesos slave that (at some point) had executors running. The external shuffle service is started externally, outside of Mesos, so Mesos does *not* know that the application exited. The external shuffle service runs on Mesos slaves all the time (per @pwendell's suggestion  [here](https://github.com/apache/spark/pull/3861#issuecomment-74950927)), and is not managed by Mesos.

We could hook that code to a shutdown hook instead of the normal path, but I want first to get to the same page regarding how it works now, and why it works the way it does.

andrewor14 replied:
Thanks for the clarification.

The reason why we saw the shuffle file not found is because we deleted the parent directory on exit, which included the shuffle directory. Your changes made these two directories siblings instead. That part makes sense.

As for who is responsible for cleaning up the shuffle files when the application exits, let's move that discussion back to the main thread.

andrewor14 commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1241,8 +1253,37 @@ private[spark] class BlockManager(
+   * shuffle files to be served by the external shuffle service. We need to delete those
+   * files when the application stops.
+   */
+  private def cleanupAllShuffleFiles() {
```
This is only best-effort. If the driver is killed forcefully we still have shuffle files lingering around.

andrewor14 commented:
@dragos Thanks for all your work and patience! I took a pass over it and I think the high level approach is reasonable. Most of my comments are relatively minor, but I have two bigger comments which I will describe here:

- How does dynamic allocation work with max cores? It won't ever scale up beyond max cores, which may not be what the user expects. There are two things we can do about this: (1) Nothing - maybe this isn't a problem after all; if the user sets max cores we should trust respect it as a hard cap, or (2) Make dynamic allocation and max cores mutually exclusive. What are your thoughts, @dragos @tnachen?

- The clean up logic here is only best-effort; if the driver is killed forcefully we won't clean up the shuffle files. By itself, this property is not much of a problem. However, a better place to detect when an application exits would be in the cluster manager itself (e.g. the Mesos master probably knows this, right?) My concern is that we'll end up with two separate code paths in the future where only one of them is needed; handling this in the cluster manager is strictly more reliable than doing so in the driver.

By the way, if you prefer, I'm completely open to separating the clean up shuffle files logic to a separate patch. This might allow us to merge the dynamic allocation part sooner. :)

dragos commented:
@andrewor14 thanks for you thorough review. I'll address your comments tomorrow (my timezone).

dragos committed:
Latest round of reviews.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36411 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36411/consoleFull) for   PR 4984 at commit [`1673c8b`](https://github.com/apache/spark/commit/1673c8b686f451e9929a95a0591fd9a6651691a5).

dragos commented:
I rebased on master and implemented the latest round of reviews. Please let me know what you think, @andrewor14 /cc @tnachen.

I'm not sure how much more time I can pour into this, so it'd be great if we can get to some sort of closure. I'm diving in [SPARK-7398](https://issues.apache.org/jira/browse/SPARK-7398) and my time in the following weeks will be limited. I can probably spend a couple hours more on this PR, and I'm happy to jump on a call to iron out the latest issues (sometimes it's so much faster to look at the code together!).

SparkQA commented:
[Test build #36411 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36411/console) for   PR 4984 at commit [`1673c8b`](https://github.com/apache/spark/commit/1673c8b686f451e9929a95a0591fd9a6651691a5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class RemoveApplication extends BlockTransferMessage `

AmplabJenkins commented:
Merged build finished. Test FAILed.

andrewor14 commented:
@dragos I think this is pretty close. The one outstanding issue is still who cleans up the shuffle files.

This patch proposes to have the driver send clean up requests to all shuffle services when the application exits. As I mentioned earlier, there are no guarantees here even if we install shutdown hooks. 
We cannot expect the driver to manage its own exit; this is really the responsibility of the cluster manager. My main objection is that for other modes like YARN and standalone mode (coming soon) we'll have two separate code paths to clean up the shuffle files, which leads to additional complexity.

I understand that the external shuffle service is started independently of Mesos, so even though the Mesos master knows when the framework exits, we still have to pass this information on to the shuffle service. There are two potential ways to make this work:
- (1) Have the shuffle service query the master for framework status periodically. When the framework has exited, we clean up the corresponding application's shuffle files. @tnachen is there enough support on the Mesos side to make this happen?
- (2) Initiate a long-running connection between the shuffle service and each driver. When the connection is closed on the other end, the shuffle service knows the corresponding application has exited.

andrewor14 commented:
Lastly, I suggested this before but I would actually recommend that we split this patch into two separate issues. The first, which is Mesos dynamic allocation (SPARK-6287), is ready for merge. The second, which is cleaning up shuffle files in Mesos when the shuffle service is used (no JIRA yet), is still being discussed. For the latter, since you have other obligations and we still can't agree on a solution yet, it's worthwhile to address it separately without blocking the first one.

@dragos How does that sound?

dragos commented:
@andrewor14 the part about cleaning up files is indeed the part that deserves some discussion. I'm very open to alternative solutions, and you summarised the challenges very well. I agree that ideally, Mesos should perform the cleanup when the framework exists, but let's wait for @tnachen for opinions.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36496 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36496/consoleFull) for   PR 4984 at commit [`9d88756`](https://github.com/apache/spark/commit/9d8875667f7aadee5d4b7bc9c1d1e16040ae93a0).

SparkQA commented:
[Test build #36496 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36496/console) for   PR 4984 at commit [`9d88756`](https://github.com/apache/spark/commit/9d8875667f7aadee5d4b7bc9c1d1e16040ae93a0).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class RemoveApplication extends BlockTransferMessage `

AmplabJenkins commented:
Merged build finished. Test PASSed.

dragos commented on core/src/main/scala/org/apache/spark/scheduler/cluster/mesos/CoarseMesosSchedulerBackend.scala:
```diff
@@ -285,4 +336,38 @@ private[spark] class CoarseMesosSchedulerBackend(
+    // executors, that means the scheduler wants to set the limit that is less than
+    // the amount of the executors that has been launched. Therefore, we take the existing
+    // amount of executors launched and deduct the executors killed as the new limit.
+    executorLimitOption = Some(Math.max(0, taskIdToSlaveId.size - pendingRemovedSlaveIds.size))
```
@tnachen the discussion is collapsed now by Github, probably because I rebased. [Here](https://github.com/apache/spark/pull/4984#discussion_r33795284) is the relevant discussion.

dragos committed:
Remove adjustment of executorLimitOption in doKillExecutors.

dragos commented:
I made the changes we agreed. I still have to rebase, that will take some time. Conflicts are not trivial, I hope to have it tomorrow.

dragos commented:
But you can still have a look and give it a go in the meantime.

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
[Test build #36802 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36802/consoleFull) for   PR 4984 at commit [`cf97666`](https://github.com/apache/spark/commit/cf97666f4ddba20eead2000cd01ba8d1b56615c6).

SparkQA commented:
[Test build #36802 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36802/console) for   PR 4984 at commit [`cf97666`](https://github.com/apache/spark/commit/cf97666f4ddba20eead2000cd01ba8d1b56615c6).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

andrewor14 commented:
In case others are following, @dragos @tnachen and I discussed offline about a way forward. To merge this feature sooner, we decided to split this patch into two separate issues: (1) mesos dynamic allocation, and (2) ensure shuffle files are cleaned up in mesos over time. This patch only tackles (1), and @tnachen will file a new JIRA for (2) and maybe fix it.

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
[Test build #36914 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36914/consoleFull) for   PR 4984 at commit [`4f30803`](https://github.com/apache/spark/commit/4f308033a55971282f3411b932bef97baefc2642).

tnachen commented:
SPARK-8873 created for cleaning up shuffle files

dragos committed:
Update tests to latest changes in core.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #36921 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36921/consoleFull) for   PR 4984 at commit [`39df8cd`](https://github.com/apache/spark/commit/39df8cd5c61828500ea89f71da5b59d43b08016e).

SparkQA commented:
[Test build #36914 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36914/console) for   PR 4984 at commit [`4f30803`](https://github.com/apache/spark/commit/4f308033a55971282f3411b932bef97baefc2642).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

dragos commented:
I've tested this on a small Mesos cluster, things look as before.

I uploaded a pre-packaged binary [here](http://downloads.typesafe.com/tmp/spark-prs/spark-1.5.0-SNAPSHOT-bin-dynAlloc020.tgz) to make it easier to test.

SparkQA commented:
[Test build #36921 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/36921/console) for   PR 4984 at commit [`39df8cd`](https://github.com/apache/spark/commit/39df8cd5c61828500ea89f71da5b59d43b08016e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

andrewor14 commented:
As discussed, this implementation currently doesn't clean up shuffle files, but this will be addressed separately by @tnachen in SPARK-8873. The dynamic allocation side LGTM so I'm merging this into master. Congratulations @dragos!

dragos commented:
:v:
